In [112]:
import pandas as pd

In [113]:
tariff_df = pd.read_csv('Tariff_20240104.csv', encoding='cp949', dtype=str)

In [163]:
# (tariff_df['품목번호'].str.len() <= 9).sum()

0

### tariff에서 필요한 feature만 select

In [114]:
new_df = tariff_df[['품목번호','관세율','관세율구분','용도세율구분','적용개시일','적용만료일']]

new_df.head()

품목번호 관세율 관세율구분 용도세율구분       적용개시일       적용만료일
0  0101211000   0     A    NaN  2024-01-01  2024-12-31
1  0101219000   8     A    NaN  2024-01-01  2024-12-31
2  0101291000   8     A    NaN  2024-01-01  2024-12-31
3  0101299000   8     A    NaN  2024-01-01  2024-12-31
4  0101300000   8     A    NaN  2024-01-01  2024-12-31

In [115]:
hscode_df = pd.read_csv('HS코드(2024).csv', dtype=str, encoding='cp949')

### hscode_df에 '2424000000' 존재

In [173]:
(hscode_df['품목번호'].str.len() <= 7).sum()

52

In [175]:
filtered_df = hscode_df[hscode_df['품목번호'].str.len() <= 9]
filtered_df

품목번호    적용시작일자    적용종료일자  \
17     01029090  20160101  20241231   
37     01059910  20160101  20241231   
51     01061950  20160101  20241231   
54     01061960  20160101  20241231   
69     01069030  20160101  20241231   
...         ...       ...       ...   
12305  96019090  20160101  20241231   
12309  96020090  20160101  20241231   
12377  96190010  20160101  20241231   
12381  96190030  20160101  20241231   
12385  96190090  20170101  20241231   

                                                   한글품목명  \
17                                                    기타   
37                                                    오리   
51                                                    여우   
54                                                    밍크   
69                                                 환형동물류   
...                                                  ...   
12305                                                 기타   
12309                                                 기타   
12377  제지용 펄프ㆍ종이ㆍ셀룰로오스워딩(cellulose wadding)ㆍ셀룰로오스섬유의 ...   
12381                                그 밖의 방직용 섬유재료로 만든 것   
12385                                                 기타   

                                                   영문품목명 HS부호내용 한국표준무역분류명  \
17                                                 Other    NaN       NaN   
37                                                 Ducks    NaN       NaN   
51                                                   Fox    NaN       NaN   
54                                                  Mink    NaN       NaN   
69                                              Annelida    NaN       NaN   
...                                                  ...    ...       ...   
12305                                              Other    NaN       NaN   
12309                                              Other    NaN       NaN   
12377  Of paper pulp, paper, cellulose wadding or web...    NaN       NaN   
12381                         Of other textile materials    NaN       NaN   
12385                                              Other    NaN       NaN   

      수량단위최대단가 중량단위최대단가 수량단위코드 중량단위코드 수출성질코드 수입성질코드 품목규격명 필수규격명 참고규격명 규격설명  \
17         NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
37         NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
51         NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
54         NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
69         NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
...        ...      ...    ...    ...    ...    ...   ...   ...   ...  ...   
12305      NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
12309      NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
12377      NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
12381      NaN      NaN    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   
12385        0        0    NaN    NaN    NaN    NaN   NaN   NaN   NaN  NaN   

      규격사항내용 성질통합분류코드 성질통합분류코드명  
17       NaN      NaN       NaN  
37       NaN      NaN       NaN  
51       NaN      NaN       NaN  
54       NaN      NaN       NaN  
69       NaN      NaN       NaN  
...      ...      ...       ...  
12305    NaN      NaN       NaN  
12309    NaN      NaN       NaN  
12377    NaN      NaN       NaN  
12381    NaN      NaN       NaN  
12385    NaN      NaN       NaN  

[1128 rows x 20 columns]

In [116]:
hscode_df[hscode_df['HS부호']== '2424000000']

HS부호    적용시작일자    적용종료일자 한글품목명            영문품목명 HS부호내용 한국표준무역분류명  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN       NaN   

     수량단위최대단가 중량단위최대단가 수량단위코드 중량단위코드 수출성질코드 수입성질코드 품목규격명 필수규격명 참고규격명 규격설명  \
2355      NaN      NaN    NaN     KG  11Z01  14Z01   NaN   NaN   NaN  NaN   

     규격사항내용  성질통합분류코드  성질통합분류코드명  
2355    NaN  14010101  (간이세율적용분)

### hscode_df 내 'HS부호' 9자리인 data 다수 확인 됨. 
#### 실제로 9자리 이하의 HS부호가 사용된다면, 현재 10자리만을 고려하는 로직과 충돌할 수 있음

In [22]:
# import requests
# import pandas as pd
# from io import StringIO
# import json

# # 요청 매개변수 설정
# params = {
#     'serviceKey': 'gGDom+BIbgsPl1jf3j7m6aMaMtBwgPBtgrenGKVs8VX2Ap9RcGYULywUHIvwxp2bI/XJ9Z7DJMqwuO8RRVMofw==',
#     'page': '1',
#     'perPage': '1000'  # 한 번의 요청으로 얻을 수 있는 최대 항목 수
# }

# # 요청 URL 설정
# url = 'https://api.odcloud.kr/api/15049722/v1/uddi:1914a5f2-6db7-4b78-9495-cd092b6e4029?page=1&perPage=1000'

# # 첫 번째 요청을 보내서 총 항목 수를 확인
# response = requests.get(url, params=params)
# data = json.loads(response.text)

In [23]:
# data

{'currentCount': 1000,
 'data': [{'HS부호': 101211000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드': '12399',
   '수출성질코드': '11Z01',
   '영문품목명': 'For farm breeding',
   '적용시작일자': '2013-01-01',
   '적용종료일자': '2022-12-31',
   '중량단위최대단가': None,
   '중량단위코드': 'KG',
   '참고규격명': None,
   '품목규격명': None,
   '필수규격명': None,
   '한국표준무역분류명': None,
   '한글품목명': '농가 사육용'},
  {'HS부호': 101219000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드': '12399',
   '수출성질코드': '11Z01',
   '영문품목명': 'Other',
   '적용시작일자': '2012-01-01',
   '적용종료일자': '2022-12-31',
   '중량단위최대단가': None,
   '중량단위코드': 'KG',
   '참고규격명': None,
   '품목규격명': None,
   '필수규격명': None,
   '한국표준무역분류명': None,
   '한글품목명': '기타'},
  {'HS부호': 101291000,
   '규격사항내용': None,
   '규격설명': None,
   '성질통합분류코드': 11020101,
   '성질통합분류코드명': '(말)',
   '수량단위최대단가': None,
   '수량단위코드': 'U',
   '수입성질코드

In [ ]:
# total_count = data['totalCount']

# # 총 페이지 수 계산
# total_pages = total_count // int(params['perPage']) + (total_count % int(params['perPage']) > 0)

# # 모든 페이지에 대해 요청을 반복
# all_data = []
# for page in range(1, total_pages + 1):
#     params['page'] = str(page)
#     response = requests.get(url, params=params)
#     df_page = pd.read_json(StringIO(response.text), dtype=str)
#     all_data.append(df_page)

# # all_data에는 이제 API가 제공하는 모든 데이터가 포함되어 있습니다.
# # 이 데이터를 하나의 DataFrame으로 합칩니다.
# df = pd.concat(all_data, ignore_index=True)

In [120]:
tariff_df.shape

(367969, 9)

### hscode_df의 feature 파악

In [23]:
hscode_df.columns 

Index(['HS부호', '적용시작일자', '적용종료일자', '한글품목명', '영문품목명', 'HS부호내용', '한국표준무역분류명',
       '수량단위최대단가', '중량단위최대단가', '수량단위코드', '중량단위코드', '수출성질코드', '수입성질코드', '품목규격명',
       '필수규격명', '참고규격명', '규격설명', '규격사항내용', '성질통합분류코드', '성질통합분류코드명'],
      dtype='object')

### hscode_df 필요 feature select

In [97]:
hscode_df = hscode_df[['HS부호', '적용시작일자', '적용종료일자', '한글품목명', '영문품목명', '수량단위코드', '중량단위코드', '수출성질코드', '수입성질코드', '성질통합분류코드', '성질통합분류코드명']]

hscode_df.head()

HS부호    적용시작일자    적용종료일자   한글품목명              영문품목명 수량단위코드 중량단위코드  \
0  0101211000  20130101  20241231  농가 사육용  For farm breeding      U     KG   
1  0101219000  20120101  20241231      기타              Other      U     KG   
2  0101291000  20120101  20241231     경주말  Horses for racing      U     KG   
3  0101299000  20120101  20241231      기타              Other      U     KG   
4  0101300000  20220101  20241231     당나귀              Asses      U     KG   

  수출성질코드 수입성질코드  성질통합분류코드  성질통합분류코드명  
0  11Z01  12399  11020101        (말)  
1  11Z01  12399  11020101        (말)  
2  11Z01  12399  11020101        (말)  
3  11Z01  12399  11020101        (말)  
4  11Z01  12399  11020190  (기타 산 동물)

In [121]:
hscode_df = hscode_df.rename(columns = {'HS부호': '품목번호'})

In [127]:
tariff_df[tariff_df['품목번호']== '2424000000']

Empty DataFrame
Columns: [품목번호, 관세율구분, 관세율, 단위당세액, 기준가격, 적용국가구분, 용도세율구분, 적용개시일, 적용만료일]
Index: []

In [126]:
hscode_df[hscode_df['품목번호']== '2424000000']

품목번호    적용시작일자    적용종료일자 한글품목명            영문품목명 HS부호내용 한국표준무역분류명  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN       NaN   

     수량단위최대단가 중량단위최대단가 수량단위코드 중량단위코드 수출성질코드 수입성질코드 품목규격명 필수규격명 참고규격명 규격설명  \
2355      NaN      NaN    NaN     KG  11Z01  14Z01   NaN   NaN   NaN  NaN   

     규격사항내용  성질통합분류코드  성질통합분류코드명  
2355    NaN  14010101  (간이세율적용분)

In [125]:
merged_df[merged_df['품목번호']== '2424000000']

Empty DataFrame
Columns: [품목번호, 관세율구분, 관세율, 단위당세액, 기준가격, 적용국가구분, 용도세율구분, 적용개시일, 적용만료일, 적용시작일자, 적용종료일자, 한글품목명, 영문품목명, HS부호내용, 한국표준무역분류명, 수량단위최대단가, 중량단위최대단가, 수량단위코드, 중량단위코드, 수출성질코드, 수입성질코드, 품목규격명, 필수규격명, 참고규격명, 규격설명, 규격사항내용, 성질통합분류코드, 성질통합분류코드명]
Index: []

[0 rows x 28 columns]

In [128]:
len(tariff_df.merge(hscode_df, on = '품목번호', how = 'left').columns)

28

In [129]:
len(hscode_df.columns) + len(tariff_df.columns) == len(tariff_df.merge(hscode_df, on = '품목번호', how = 'left').columns) +1

True

In [146]:
len(hscode_df)

12422

In [147]:
# mergedHsTariff = tariff_df.merge(hscode_df, on = '품목번호', how = 'left')
mergedHsTariff = hscode_df.merge(tariff_df, on = '품목번호', how = 'left')

In [148]:
mergedHsTariff = mergedHsTariff.drop(columns = ['기준가격'])

In [151]:
mergedHsTariff.head()

품목번호    적용시작일자    적용종료일자   한글품목명              영문품목명 HS부호내용 한국표준무역분류명  \
0  0101211000  20130101  20241231  농가 사육용  For farm breeding    NaN       NaN   
1  0101211000  20130101  20241231  농가 사육용  For farm breeding    NaN       NaN   
2  0101211000  20130101  20241231  농가 사육용  For farm breeding    NaN       NaN   
3  0101211000  20130101  20241231  농가 사육용  For farm breeding    NaN       NaN   
4  0101211000  20130101  20241231  농가 사육용  For farm breeding    NaN       NaN   

  수량단위최대단가 중량단위최대단가 수량단위코드  ... 규격사항내용  성질통합분류코드 성질통합분류코드명 관세율구분   관세율 단위당세액  \
0      NaN      NaN      U  ...    NaN  11020101       (말)     A     0   NaN   
1      NaN      NaN      U  ...    NaN  11020101       (말)     C  13.1   NaN   
2      NaN      NaN      U  ...    NaN  11020101       (말)  FAS1     0   NaN   
3      NaN      NaN      U  ...    NaN  11020101       (말)  FAU1     0   NaN   
4      NaN      NaN      U  ...    NaN  11020101       (말)  FCA1     0   NaN   

  적용국가구분 용도세율구분       적용개시일       적용만료일  
0      1    NaN  2024-01-01  2024-12-31  
1      1    NaN  2024-01-01  2024-12-31  
2      2    NaN  2024-01-01  2024-12-31  
3      2    NaN  2024-01-01  2024-12-31  
4      2    NaN  2024-01-01  2024-12-31  

[5 rows x 27 columns]

In [156]:
cols = ['품목번호', '한글품목명', '영문품목명','관세율','관세율구분', '성질통합분류코드명'] + [col for col in mergedHsTariff.columns if col not in ['품목번호', '한글품목명', '영문품목명','관세율' ,'관세율구분','성질통합분류코드명']]
mergedHsTariff = mergedHsTariff[cols]

In [157]:
mergedHsTariff.head()

품목번호   한글품목명              영문품목명   관세율 관세율구분 성질통합분류코드명    적용시작일자  \
0  0101211000  농가 사육용  For farm breeding     0     A       (말)  20130101   
1  0101211000  농가 사육용  For farm breeding  13.1     C       (말)  20130101   
2  0101211000  농가 사육용  For farm breeding     0  FAS1       (말)  20130101   
3  0101211000  농가 사육용  For farm breeding     0  FAU1       (말)  20130101   
4  0101211000  농가 사육용  For farm breeding     0  FCA1       (말)  20130101   

     적용종료일자 HS부호내용 한국표준무역분류명  ... 필수규격명 참고규격명 규격설명 규격사항내용  성질통합분류코드 단위당세액  \
0  20241231    NaN       NaN  ...   NaN   NaN  NaN    NaN  11020101   NaN   
1  20241231    NaN       NaN  ...   NaN   NaN  NaN    NaN  11020101   NaN   
2  20241231    NaN       NaN  ...   NaN   NaN  NaN    NaN  11020101   NaN   
3  20241231    NaN       NaN  ...   NaN   NaN  NaN    NaN  11020101   NaN   
4  20241231    NaN       NaN  ...   NaN   NaN  NaN    NaN  11020101   NaN   

  적용국가구분 용도세율구분       적용개시일       적용만료일  
0      1    NaN  2024-01-01  2024-12-31  
1      1    NaN  2024-01-01  2024-12-31  
2      2    NaN  2024-01-01  2024-12-31  
3      2    NaN  2024-01-01  2024-12-31  
4      2    NaN  2024-01-01  2024-12-31  

[5 rows x 27 columns]

In [183]:
#len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 10])
len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 9])

170

In [144]:
len(mergedHsTariff[mergedHsTariff['품목번호'].str.len() == 10])

367970

In [108]:
len(mergedHsTariff)

367969

In [109]:
unique_count = mergedHsTariff['품목번호'].nunique()
print(unique_count)

11293


In [110]:
len(hscode_df) -unique_count # 1129개가 어디서 나왔는가?

1129

In [145]:
hscode_df[hscode_df['품목번호']== '2424000000']

품목번호    적용시작일자    적용종료일자 한글품목명            영문품목명 HS부호내용 한국표준무역분류명  \
2355  2424000000  20020101  20241231  이사화물  household goods    NaN       NaN   

     수량단위최대단가 중량단위최대단가 수량단위코드 중량단위코드 수출성질코드 수입성질코드 품목규격명 필수규격명 참고규격명 규격설명  \
2355      NaN      NaN    NaN     KG  11Z01  14Z01   NaN   NaN   NaN  NaN   

     규격사항내용  성질통합분류코드  성질통합분류코드명  
2355    NaN  14010101  (간이세율적용분)

HSK는 세계관세기구가 제정한 6단위 코드로 구성된 국제 통일 상품분류체계(HS)를 바탕으로 우리나라의 필요에 따라 품목을 세분화한 10단위 품목분류표인데, 이번 개정으로 HSK상 품목이 12,242개에서 11,293개(신설 341개, 삭제 1,290개)로 변경되었습니다.

In [56]:

len(hscode_df[hscode_df['품목번호'].str.len() <= 9]) # 1개가 어디서 왔을까?

1128

In [60]:
len(hscode_df['품목번호'].unique())

12422

In [68]:
unique_merged = set(mergedHsTariff['품목번호'].unique())
unique_hscode = set(hscode_df['품목번호'].unique())

non_duplicates = set(unique_merged.symmetric_difference(unique_hscode))
len(non_duplicates)

1129

In [65]:
missingValue =list(non_duplicates.symmetric_difference(unique_hscode))

In [67]:
len(missingValue)

11293

In [69]:
non_duplicates_set = set(non_duplicates)
unique_hscode = set(hscode_df['품목번호'].unique())

difference = list(non_duplicates_set - unique_hscode)

In [71]:
unique_hscode = set(hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].unique()) # 1128
unique_merged = set(hscode_df['품목번호'] - mergedHsTariff['품목번호'].unique())

only_in_hscode = list(unique_hscode - unique_merged)

In [75]:
unique_hscode = set(hscode_df['품목번호'].unique())
unique_merged = set(mergedHsTariff['품목번호'].unique())

only_in_hscode = list(unique_hscode - unique_merged)

In [79]:
only_in_hscode

1129

In [80]:
hscode_df[hscode_df['품목번호'].str.len() <= 9]

1128

In [81]:
filtered_hscode = set(hscode_df[hscode_df['품목번호'].str.len() <= 9]['품목번호'].unique())
only_in_hscode_set = set(only_in_hscode)

non_overlap = list(filtered_hscode.symmetric_difference(only_in_hscode_set))

In [82]:
non_overlap #찾았다.

['2424000000']

In [91]:
hscode_df['품목번호'=='2424000000']

KeyError: False

In [88]:
mergedHsTariff[mergedHsTariff['품목번호'=='2424000000']]

KeyError: False